File: Memento/Ruby/EditorHistoryCaretaker.rb
require_relative "WorkspaceState"

class EditorHistoryCaretaker
    def initialize
        @states = []
    end

    def add(state)
        @states.push(state)
    end

    def get(index)
        if @states.length < 1
            return nil
        end

        if index < @states.length
            @states[index]
        else
            @states[@states.length - 1]
        end
    end
end

File: Memento/Python 3/EditorHistoryCaretaker.py
from WorkspaceState import WorkspaceState

class EditorHistoryCaretaker:
  def __init__(self):
    self.states = []
  
  def add(self, state):
    self.states.append(state)
  
  def get(self, index):
    if len(self.states) < 1:
      return None
    
    if index < len(self.states):
      return self.states[index]
    else:
      return self.states[-1]